hive/agents/qnets/rnn.py

```diff
@@ -0,0 +1,138 @@
+                layers and mlp layers.
+            channels (list): The size of output channel for each convolutional layer.
+            mlp_layers (list): The number of neurons for each mlp layer after the
+                convolutional layers.
+            kernel_sizes (list | int): The kernel size for each convolutional layer
+            strides (list | int): The stride used for each convolutional layer.
+            paddings (list | int): The size of the padding used for each convolutional
+                layer.
+            normalization_factor (float | int): What the input is divided by before
+                the forward pass of the network.
+            noisy (bool): Whether the MLP part of the network will use
+                :py:class:`~hive.agents.qnets.noisy_linear.NoisyLinear` layers or
+                :py:class:`torch.nn.Linear` layers.
+            std_init (float): The range for the initialization of the standard
+                deviation of the weights in
+                :py:class:`~hive.agents.qnets.noisy_linear.NoisyLinear`.
+        """
+        super().__init__()
+        self._normalization_factor = normalization_factor
+        if channels is not None:
+            if isinstance(kernel_sizes, int):
+                kernel_sizes = [kernel_sizes] * len(channels)
+            if isinstance(strides, int):
+                strides = [strides] * len(channels)
+            if isinstance(paddings, int):
+                paddings = [paddings] * len(channels)
+
+            if not all(
+                len(x) == len(channels) for x in [kernel_sizes, strides, paddings]
+            ):
+                raise ValueError("The lengths of the parameter lists must be the same")
+
+            # Convolutional Layers
+            channels.insert(0, in_dim[0])
+            conv_seq = []
+            for i in range(0, len(channels) - 1):
+                conv_seq.append(
+                    torch.nn.Conv2d(
+                        in_channels=channels[i],
+                        out_channels=channels[i + 1],
+                        kernel_size=kernel_sizes[i],
+                        stride=strides[i],
+                        padding=paddings[i],
+                    )
+                )
+                conv_seq.append(torch.nn.ReLU())
+            self.conv = torch.nn.Sequential(*conv_seq)
+        else:
+            self.conv = torch.nn.Identity()
+
+        # RNN Layers
+        conv_output_size = calculate_output_dim(self.conv, in_dim)
+        self.rnn = sequence_fn(
+            rnn_input_size=np.prod(conv_output_size),
+        )
+
+        if mlp_layers is not None:
+            # MLP Layers
+            self.mlp = MLPNetwork(
+                sequence_fn.keywords["rnn_hidden_size"],
+                mlp_layers,
+                noisy=noisy,
+                std_init=std_init,
+            )
+        else:
+            self.mlp = nn.Identity()
+
+    def forward(self, x, hidden_state=None):
+        if len(x.shape) == 4:
```

Comment:
can be cleaned if we have separate call functions to act and update, as suggested above.

Reply:
If we do so we would have two `forward()` functions with a big chunk of shared code. Also we need the hidden state to compute the Q values for both act and update. There is no extra computation involved to compute the new hidden states. Thus for now it's fine if we use the same function for both act and updatae.

Reply:
What you can do is have the act function modify the input and then pass it to forward? Just looks like a cleaner way to separate two use cases instead of having an if condition.

Reply:
done